adapters/repos/db/shard.go

```diff
@@ -985,23 +1033,72 @@ func (s *Shard) Shutdown(ctx context.Context) error {
 		}
 	}
 
-	// unregister all callbacks at once, in parallel
-	if err = cyclemanager.NewCombinedCallbackCtrl(0, s.index.logger,
-		s.cycleCallbacks.compactionCallbacksCtrl,
-		s.cycleCallbacks.flushCallbacksCtrl,
-		s.cycleCallbacks.vectorCombinedCallbacksCtrl,
-		s.cycleCallbacks.geoPropsCombinedCallbacksCtrl,
-	).Unregister(ctx); err != nil {
-		return err
-	}
-
 	if err = s.store.Shutdown(ctx); err != nil {
 		return errors.Wrap(err, "stop lsmkv store")
 	}
 
 	return nil
 }
 
+func (s *Shard) preventShutdown() (release func(), err error) {
+	s.shutdownLock.RLock()
+	defer s.shutdownLock.RUnlock()
+
+	if s.shut {
+		return func() {}, fmt.Errorf("shard %q already shut or dropped", s.name)
+	}
+
+	s.inUseCounter.Add(1)
+	return func() { s.inUseCounter.Add(-1) }, nil
+}
+
+func (s *Shard) waitForShutdown(ctx context.Context) error {
+	checkInterval := 50 * time.Millisecond
+	timeout := 30 * time.Second
+
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	if eligible, err := s.checkEligibleForShutdown(); err != nil {
+		return err
+	} else if !eligible {
+		ticker := time.NewTicker(checkInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return fmt.Errorf("Shard::proceedWithShutdown: %w", ctx.Err())
```

Comment:
**thought:**  what do you think if we log if `ctx.Err()` != nil and always return nil, because if the ctx got cancelled for reasons like was finished etc. it's still not an error

Reply:
As I understand method with context as an argument should respect context's timeout. Shutdown gets context, so if it reaches timeout it should fail and return error. Here timeout can be reached due to shard being in use for too long before even starting actual shutdown.

Reply:
resolve to merge